### PR TITLE
feat: Support ARM endpoint with non-trusted certificate

### LINKED ISSUE #280: Support ARM endpoint with non-trusted certificate
**Is this a request for help?**:
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
Feature Request
---

**What version of aks-engine?**:
the latest version
---

Customer will provide the certificate to binding with external endpoint. e.g ARM endpoint.

The certificate could be non-trusted on the Linux Machine since customer could use cooperated signed certificate to deploy azure stack.

Right now, the k8s controller will be not be able to communicate back to ARM endpoint with non-trusted certificate. 

AKS engine would need trusted the certificate and mount the certificate to the k8s controllers container.

### PR DESCRIPTION
**Pull request by honcao**
**Reason for Change**:

feat: Support ARM endpoint with non-trusted certificate

**Issue Fixed**:

Fixes #280 

**Requirements**:

- [X] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:
Customer will provide the certificate to binding with external endpoint. e.g ARM endpoint.

The certificate could be non-trusted on the Linux Machine since customer could use cooperated signed certificate to deploy azure stack.

Right now, the k8s controller will be not be able to communicate back to ARM endpoint with non-trusted certificate.

AKS engine would need trusted the certificate and mount the certificate to the k8s controllers container.

### CODE REVIEW

**Comment by jackfrancis** [comment]
@honcao, let's do this, to avoid adding add'l ARM deployment payload unnecessarily to Azure Stack-only scenarios. See:

```
- path: /opt/azure/containers/provision_source.sh
  permissions: "0744"
  encoding: gzip
  owner: root
  content: !!binary |
    {{WrapAsVariable "provisionSource"}}
```

☝️ in the `kubernetesagentcustomdata.yml` file (cloud-init config). Basically:

1. Let's move all Azure Stack-specific shell functionality into a new shell script
2. Conditionally add it to the cloud-init configuration (the appropriate *customdata.yml files under parts/)
3) Conditionally source this new file (or files) into the `kubernetescustomscript.sh` implementation *if* the provision is operating in an Azure Stack context

Does the above make sense?

**Comment by mboersma** [comment]
@honcao I rebased and force-pushed this PR to reconcile it with #546 in master. It's no longer necessary to commit the generated files. Sorry about that!

**Comment by honcao** [comment]
@jackfrancis , I will make the changes to implement your proposal

File: parts/k8s/kubernetesconfigscustomcloud.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+function ensureCertificates()
```
Comment (jackfrancis) [inline parts/k8s/kubernetesconfigscustomcloud.sh]: nit: let's use the `ensureCertificates() {` function definition pattern (like you have below for `configureK8sCustomCloud`)

File: parts/k8s/kubernetescustomscript.sh [hunk 2]
```diff
@@ -24,6 +24,12 @@ config_script=/opt/azure/containers/provision_configs.sh
 wait_for_file 3600 1 $config_script || exit $ERR_FILE_WATCH_TIMEOUT
 source $config_script
 
+if [[ "${TARGET_ENVIRONMENT,,}" == "azurestackcloud"  ]]; then
```
Comment (jackfrancis) [inline parts/k8s/kubernetescustomscript.sh]: Let's add an `AZURE_STACK_ENV="azurestackcloud" # Azure Stack target env string` const-like thing to the `kubernetesprovisionsource.sh` file (we source it in above in the `source $script_lib` statement).

And then refer to `$AZURE_STACK_ENV` here whenever we want to compare a string to `"azurestackcloud"`.

File: parts/k8s/kubernetescustomscript.sh [hunk 3]
```diff
@@ -24,6 +24,12 @@ config_script=/opt/azure/containers/provision_configs.sh
 wait_for_file 3600 1 $config_script || exit $ERR_FILE_WATCH_TIMEOUT
 source $config_script
 
+if [[ "${TARGET_ENVIRONMENT,,}" == "azurestackcloud"  ]]; then 
+config_script_custom_cloud=/opt/azure/containers/provision_configs_custom_cloud.sh
```
Comment (jackfrancis) [inline parts/k8s/kubernetescustomscript.sh]: nit: indent 4 spaces inside `if` blocks

File: parts/k8s/kubernetescustomscript.sh [hunk 4]
```diff
@@ -124,6 +130,11 @@ elif [[ "$CONTAINER_RUNTIME" == "kata-containers" ]]; then
 fi
 
 configureK8s
+
+if [[ "${TARGET_ENVIRONMENT,,}" == "azurestackcloud"  ]]; then 
+configureK8sCustomCloud
```
Comment (jackfrancis) [inline parts/k8s/kubernetescustomscript.sh]: nit: indent 4 spaces inside `if` blocks

File: parts/k8s/kubernetescustomscript.sh (hunk 2 above)
Comment (honcao) [reply parts/k8s/kubernetescustomscript.sh]: Fixed.

File: parts/k8s/kubernetesconfigscustomcloud.sh [hunk 5]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+AZURE_STACK_ENV="azurestackcloud"
```
Comment (jackfrancis) [inline parts/k8s/kubernetesconfigscustomcloud.sh]: Move this to `kubernetesprovisionsource.sh`

File: parts/k8s/kubernetesconfigscustomcloud.sh [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        fi
+
+        # ensureCertificates will be retried if the exit code is not 0
+        curl $AZURESTACK_RESOURCE_METADATA_ENDPOINT
+        exit $?
+    fi
+}
+
+configureK8sCustomCloud() {
+    if [[ "${TARGET_ENVIRONMENT,,}" == "${AZURE_STACK_ENV}"  ]]; then
```
Comment (jackfrancis) [inline parts/k8s/kubernetesconfigscustomcloud.sh]: We don't need this `if` check, as the same `if` check is around the function invocation (where it should be).

File: parts/k8s/kubernetesconfigscustomcloud.sh [hunk 7]
```diff
@@ -0,0 +1,72 @@
+#!/bin/bash
+AZURE_STACK_ENV="azurestackcloud"
+function ensureCertificates() {
+    if [[ "${TARGET_ENVIRONMENT,,}" == "${AZURE_STACK_ENV}" ]]; then
```
Comment (jackfrancis) [inline parts/k8s/kubernetesconfigscustomcloud.sh]: We don't need this `if` check, as the only caller of this function is `configureK8sCustomCloud`, which is already guarded by the same `if` check.

File: parts/k8s/kubernetesconfigscustomcloud.sh [hunk 8]
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+function ensureCertificates() {
```
Comment (jackfrancis) [inline parts/k8s/kubernetesconfigscustomcloud.sh]: replace `function ensureCertificates() {` with `ensureCertificates() {`

File: parts/k8s/kubernetesagentcustomdata.yml [hunk 9]
```diff
@@ -28,6 +28,15 @@ write_files:
   owner: root
   content: !!binary |
     {{WrapAsVariable "provisionConfigs"}}
+
```
Comment (jackfrancis) [inline parts/k8s/kubernetesagentcustomdata.yml]: Are we sure we need this custom controller manager scaffolding on agent nodes? For regular (non-Azure Stack) scenarios, the controller manager implementation is only on master nodes.

File: parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  name: kube-controller-manager
+  namespace: kube-system
+  labels:
+    tier: control-plane
+    component: kube-controller-manager
+spec:
+  hostNetwork: true
+  containers:
+    - name: kube-controller-manager
+      image: <img>
```
Comment (jackfrancis) [inline parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml]: I don't see anything that will properly inject a value into ``

File: parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+      imagePullPolicy: IfNotPresent
+      command: ["/hyperkube", "controller-manager"]
+      args: [<args>]
```
Comment (jackfrancis) [inline parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml]: I don't see anything that will properly inject a value into `<args>`

File: parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml (hunk 10 above)
Comment (honcao) [reply parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml]: src/github.com/Azure/aks-engine/parts/k8s/kubernetesmastercustomdata.yml
line 308 to 311
/opt/azure/containers/kubelet.sh will replace the

File: parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml (hunk 11 above)
Comment (honcao) [reply parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml]: src/github.com/Azure/aks-engine/parts/k8s/kubernetesmastercustomdata.yml
line 319
/opt/azure/containers/kubelet.sh will replace the <args>

File: parts/k8s/kubernetesagentcustomdata.yml (hunk 9 above)
Comment (honcao) [reply parts/k8s/kubernetesagentcustomdata.yml]: It is also true that controller manager implementation is only on master nodes.
I updated ensureCertificates to check the existence kube controller file before update it.
The logic now are:
 

- on Master trust the certificate and update the kube controller config 

- on Agent, only trust the certificate

File: parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml (hunk 10 above)
Comment (jackfrancis) [reply parts/k8s/manifests/kubernetesmaster-kube-controller-manager-custom.yaml]: Got it, you are re-using the `kube-controller-manager.yaml` filename for Azure Stack scenarios.

**Pull request merged**